src/main/java/picard/sam/SamErrorMetric/CollectSamErrorMetrics.java

```diff
@@ -328,30 +416,205 @@ private void writeMetricsFileForAggregator(final BaseErrorAggregation locusAggre
-        while (vcfIterator.hasNext() && (vcfIterator.peek().isFiltered() ||
-                CompareVariantContextToLocus(sequenceDictionary, vcfIterator.peek(), locus) < 0)) {
-            vcfIterator.next();
+    private static boolean checkLocus(final VCFFileReader vcfFileReader, final Locus locus, final SAMSequenceDictionary sequenceDictionary) {
+
+        boolean overlaps = false;
+        final Interval queryInterval = locusToInterval(locus, sequenceDictionary);
+
+        if ( queryInterval != null ) {
+
+            try ( final CloseableIterator<VariantContext> vcfIterator = vcfFileReader.query(queryInterval) ) {
+
+                overlaps = true;
+
+                boolean allFiltered = true;
+
+                while ( vcfIterator.hasNext() ) {
+                    final VariantContext vcf = vcfIterator.next();
+                    if ( vcf.isFiltered() ) {
+                        continue;
+                    }
+                    allFiltered = false;
+                }
+
+                if ( allFiltered ) {
+                    overlaps = false;
+                }
+            }
+        }
+
+        return overlaps;
+    }
+
+    /**
+     * Converts the given locus into an interval using the given sequenceDictionary.
+     * @param locus The {@link Locus} to convert.
+     * @param sequenceDictionary The {@link SAMSequenceDictionary} to use to convert the given {@code locus}.
+     * @return An {@link Interval} representing the given {@code locus} or {@code null} if it cannot be converted.
+     */
+    private static Interval locusToInterval(final Locus locus, final SAMSequenceDictionary sequenceDictionary) {
+        final SAMSequenceRecord samSequenceRecord = sequenceDictionary.getSequence( locus.getSequenceIndex() );
+        if ( samSequenceRecord == null ) {
+            return null;
+        }
+
+        return new Interval(samSequenceRecord.getSequenceName(), locus.getPosition(), locus.getPosition());
+    }
+
+    /**
+     * Map of previously seen deletion records, associated with the locus they have been last seen
+     */
+    private HashMap<SAMRecord, SamLocusIterator.LocusInfo> previouslySeenDeletions = new HashMap<>();
+
+    /**
+     * Current locus for taking care of deleting deletion records from the above map
+     */
+    private SamLocusIterator.LocusInfo currentLocus = null;
+
+    /**
+     * Checks if the same record has been seen at the previous locus already, thereby determining
+     * whether or not a deletion has already been processed. Note that calling this method will
+     * signal that the record is processed at this location.
+     * @param deletionRaO The RecordAndOffset to be checked
+     * @param locusInfo The LocusInfo to determine the current locus
+     * @return True, if the record has been seen at the previous locus.
+     */
+    @VisibleForTesting
+    protected boolean hasDeletionBeenProcessed(SamLocusIterator.RecordAndOffset deletionRaO, SamLocusIterator.LocusInfo locusInfo) {
```

Comment:
methods called "has"- something shouldn't have side-effects...meaning that one should be able to call a "has" function twice and get the same result. but this method has side effects...so it should be called "process"-somthing and return true or false depending on whether it has been already processed.